src/app/contexts/Ledger/use-ledger.ts

```diff
@@ -0,0 +1,182 @@
+
+export type LedgerData = {
+	address: string;
+	balance: BigNumber;
+	index: number;
+	isNew?: boolean;
+};
+
+export const useLedger = (transport: typeof Transport) => {
+	const { env, persist } = useEnvironmentContext();
+	const [state, dispatch] = useReducer(ledgerStateReducer, defaultLedgerState);
+	const abortRetryRef = useRef<boolean>(false);
+
+	const { isBusy, isConnected, error } = state;
+
+	const listenDevice = useCallback(
+		() =>
+			transport.listen({
+				// @ts-ignore
+				next: ({ type, descriptor, deviceModel }) => {
+					if (type === "add") {
+						dispatch({ type: "add", path: descriptor, id: deviceModel?.id || "nanoS" });
+						return;
+					}
+					dispatch({ type: "remove" });
+				},
+				error: (e) => dispatch({ type: "failed", message: e.message }),
+				complete: () => void 0,
+			}),
+		[transport],
+	);
+
+	const importLedgerWallets = useCallback(
+		async (wallets: LedgerData[], coin: Coins.Coin, profile: Profile) => {
+			for (const { address, index } of wallets) {
+				const wallet = await profile
+					.wallets()
+					.importByAddress(address, coin.network().coin(), coin.network().id());
+				wallet.data().set(WalletFlag.LedgerIndex, index);
+			}
+			await persist();
+		},
+		[persist],
+	);
+
+	const connect = useCallback(
+		async (
+			coin: string,
+			network: string,
+			retryOptions: retry.Options = { retries: 50, randomize: false, factor: 1 },
+		) => {
+			dispatch({ type: "waiting" });
+			abortRetryRef.current = false;
+
+			const instance = await env.coin(coin, network);
+
+			try {
+				const slip44 = instance.config().get<number>("network.crypto.slip44");
+
+				const connectFn: retry.RetryFunction<void> = async (bail, attempts) => {
+					if (abortRetryRef.current && attempts > 1) {
+						bail(new Error("User aborted"));
+					}
+
+					await instance.ledger().connect(transport);
+					// Ensure that the app is accessible
+					await instance.ledger().getPublicKey(formatDerivationPath(slip44, 0));
+				};
+
+				await retry(connectFn, retryOptions);
+				dispatch({ type: "connected" });
+			} catch (connectError) {
+				instance.ledger().disconnect();
+				dispatch({ type: "failed", message: connectError.message });
+				throw connectError;
+			}
+		},
+		[transport, env],
+	);
+
+	const disconnect = useCallback(async (coin: Coins.Coin) => {
+		await coin.ledger().disconnect();
+		dispatch({ type: "disconnected" });
+	}, []);
+
+	const scanWallets = useCallback(
+		async (coin: string, network: string, profile: Profile, onChange?: (wallet: LedgerData) => void) => {
+			const wallets: LedgerData[] = [];
+
+			try {
+				await connect(coin, network);
+
+				const instance = await env.coin(coin, network);
+				const slip44 = instance.config().get<number>("network.crypto.slip44");
+
+				dispatch({ type: "busy" });
+
+				let hasMore = true;
+				let cursor = 0;
+
+				while (hasMore) {
+					const path = formatDerivationPath(slip44, cursor);
```

Comment:
This whole process of preparing the path, publicKey, and address should ideally be in the SDK as an `importFromLedger` method. The wallet has way too much knowledge about the specifics of how the ledger data is used to create a wallet from it.

Reply:
Agreed